### PR TITLE
Added download and delete lessons and chapter support

### PR DESCRIPTION
**Pull request by ALOK-TV**

### CODE REVIEW

File: public/assets/icons/LessonCard.tsx [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+  const [currentCourse, setCurrentCourse] = useState<Course>();
+
+  const hideImg = (event: any) => {
+    setShowImage(false);
+  };
+  // const subjectCode = lesson.chapter.course.id;
+  useEffect(() => {
+    // getSubject();
+    getCurrentCourse();
+  }, [lesson]);
```
Comment (bhanugummana) [inline public/assets/icons/LessonCard.tsx:67]: why do you need to get getCurrentCourse when ever lesson is changed??

File: public/assets/icons/LessonCard.tsx [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+  };
+
+  // const lessonCardColor =
+  //   LESSON_CARD_COLORS[Math.floor(Math.random() * LESSON_CARD_COLORS.length)];
+
+  const [lessonCardColor, setLessonCardColor] = useState("");
+
+  useEffect(() => {
+    setLessonCardColor(
+      LESSON_CARD_COLORS[Math.floor(Math.random() * LESSON_CARD_COLORS.length)]
```
Comment (bhanugummana) [inline public/assets/icons/LessonCard.tsx:111]: will this change to new color every time we side??

File: public/assets/icons/LessonCard.tsx [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    );
+  }, []);
+
+  return (
+    <div>
+      <IonCard
+        id="lesson-card"
+        style={{
+          width: width,
+          height: "auto",
```
Comment (bhanugummana) [inline public/assets/icons/LessonCard.tsx:121]: set this in css file

File: public/assets/icons/LessonCard.tsx [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+            // }
+          } else {
+            console.log(lesson?.title, "lesson is locked");
+          }
+        }}
+        // disabled={!isUnlocked}
+      >
+        <div
+          style={{
+            display: "grid",
```
Comment (bhanugummana) [inline public/assets/icons/LessonCard.tsx:167]: set this styles in css file

File: public/assets/icons/LessonCard.tsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+              parmas,
+              Lesson.toJson(lesson)
+            );
+            history.push(PAGES.GAME + parmas, {
+              url: "chimple-lib/index.html" + parmas,
+              lessonId: lesson.id,
+              courseDocId: course?.docId ?? lesson?.assignment?.course?.id,
+              course: JSON.stringify(Course.toJson(currentCourse!)),
+              lesson: JSON.stringify(Lesson.toJson(lesson)),
+              from: history.location.pathname + "?continue=true",
```
Comment (bhanugummana) [inline public/assets/icons/LessonCard.tsx:156]: take continue as a const from @mujafar7 PR

File: public/assets/icons/LessonCard.tsx [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            display: "grid",
+          }}
+        >
+          <div
+            style={{
+              background: lessonCardColor,
+              borderRadius: "7vh",
+              width: width,
+              height: height,
+              display: "grid",
```
Comment (bhanugummana) [inline public/assets/icons/LessonCard.tsx:176]: set this styles in css file

File: public/assets/icons/LessonCard.tsx [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+                  "https://firebasestorage.googleapis.com/v0/b/cuba-stage.appspot.com/o/lesson_thumbnails%2FlessonCaredPattern%2FChallengePattern.png?alt=media&token=be64aec1-f70f-43c3-95de-fd4b1afe5806"
+                }
+              />
+            </div>
+
+            <div id="lesson-card-image">
+              <SelectIconImage
+                localSrc={
+                  "courses/" +
+                  lesson.cocosSubjectCode +
```
Comment (bhanugummana) [inline public/assets/icons/LessonCard.tsx:220]: go with the current course code, not with the cocosSubjectCode

File: src/components/DownloadLesson.tsx [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    function handleOfflineEvent() {
+      setOnline(false);
+    }
+
+    window.addEventListener("online", handleOnlineEvent);
+    window.addEventListener("offline", handleOfflineEvent);
+
+    if (isStored(lessonID, "lessonItems")) {
```
Comment (bhanugummana) [inline src/components/DownloadLesson.tsx:36]: use constants for lessonItems and chapterItems

File: src/components/DownloadLesson.tsx [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+      setShowIcon(false);
+    }
+
+    if (chapters && isStored(chapters.id, "chapterItems")) {
+      setShowIcon(false);
+    }
+  }, [lessonID, chapters]);
```
Comment (bhanugummana) [inline src/components/DownloadLesson.tsx:43]: do we really need these dependencies??

File: src/components/DownloadLesson.tsx [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+  const storeId = (id: string, storageKey: string) => {
+    const storedItems = JSON.parse(localStorage.getItem(storageKey) || "[]");
+    const updatedItems = [...storedItems, id];
+    localStorage.setItem(storageKey, JSON.stringify(updatedItems));
+  };
+  const removeFromStorage = (id: string, storageKey: string) => {
+    const storedItems = JSON.parse(localStorage.getItem(storageKey) || "[]");
+    const updatedItems = storedItems.filter((item: string) => item !== id);
+    localStorage.setItem(storageKey, JSON.stringify(updatedItems));
```
Comment (bhanugummana) [inline src/components/DownloadLesson.tsx:58]: check if this updatedItems is empty or null

File: src/components/DownloadLesson.tsx [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    console.log("lesson is deleted");
+  };
+  const fetchData = async () => {
+    if (loading) return;
+    setLoading(true);
+
+    setTimeout(async () => {
+      if (!online) {
+        showSnackbar(
+          `Device is offline. Cannot download  ${
```
Comment (bhanugummana) [inline src/components/DownloadLesson.tsx:68]: use i18n for this text

File: src/components/DownloadLesson.tsx [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    }, 3000);
+  };
+
+  const handleDelete = async () => {
+    if (loading) return;
+    setLoading(true);
+    setTimeout(async () => {
+      if (chapters) {
+        removeFromStorage(chapters.id, "chapterItems");
+        await deleteLessons(chapters.id);
```
Comment (bhanugummana) [inline src/components/DownloadLesson.tsx:108]: why are you doing the deleteLesson and other things in the setTimeout??

File: src/components/DownloadLesson.tsx [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+    if (!dow) {
+      return;
+    }
+    console.log("downloaded ", dow);
+  }
+
+  const showSnackbar = (message, type) => {
+    switch (type) {
+      case "success":
```
Comment (bhanugummana) [inline src/components/DownloadLesson.tsx:158]: use this types as enum

File: src/utility/util.ts [hunk 14]
```diff
@@ -357,6 +363,52 @@ export class Util {
...
+        recursive: true,
+      });
+      console.log("lessonIDD", lessonId);
+    } catch (error) {
+      console.log("Error deleting lesson:", error);
+    }
+    return true;
+  }
+
+  public static async checkDownloadedLessons() {
```
Comment (bhanugummana) [inline src/utility/util.ts:384]: Don't check it every time the user opens the app; check after some time interval

File: src/utility/util.ts [hunk 15]
```diff
@@ -357,6 +363,52 @@ export class Util {
...
+      await Filesystem.rmdir({
+        path: `${lessonId}`,
+        directory: Directory.External,
+        recursive: true,
+      });
+      console.log("lessonIDD", lessonId);
+    } catch (error) {
+      console.log("Error deleting lesson:", error);
+    }
+    return true;
```
Comment (bhanugummana) [inline src/utility/util.ts:381]: why is it returning true even whey delete failed

File: src/utility/util.ts [hunk 16]
```diff
@@ -357,6 +363,52 @@ export class Util {
...
+          localStorage.getItem(storageKey) || "[]"
+        );
+        const updatedItems = [...storedItems, id];
+        localStorage.setItem(storageKey, JSON.stringify(updatedItems));
+      };
+      const folderNamesArray: string[] = [];
+      const folderNames = contents.files;
+      for (let i = 0; i < folderNames.length; i++) {
+        console.log("Processing folder:", folderNames[i].name);
+        storeId(folderNames[i].name, "lessonItems");
```
Comment (bhanugummana) [inline src/utility/util.ts:403]: try to update the local storage all at once not on every loop

File: src/utility/util.ts [hunk 17]
```diff
@@ -357,6 +363,52 @@ export class Util {
...
+      const contents = await Filesystem.readdir({
+        path: "",
+        directory: Directory.External,
+      });
+      const storeId = (id: string, storageKey: string) => {
+        const storedItems = JSON.parse(
+          localStorage.getItem(storageKey) || "[]"
+        );
+        const updatedItems = [...storedItems, id];
+        localStorage.setItem(storageKey, JSON.stringify(updatedItems));
```
Comment (bhanugummana) [inline src/utility/util.ts:397]: how are you removing the deleted lessonId from the localStorage

File: src/utility/util.ts [hunk 18]
```diff
@@ -357,6 +363,52 @@ export class Util {
...
+      if (!Capacitor.isNativePlatform()) return null;
+
+      const contents = await Filesystem.readdir({
+        path: "",
+        directory: Directory.External,
+      });
+      const storeId = (id: string, storageKey: string) => {
+        const storedItems = JSON.parse(
+          localStorage.getItem(storageKey) || "[]"
+        );
```
Comment (bhanugummana) [inline src/utility/util.ts:395]: how are you handling the localstorage for chapters

**Review by bhanugummana — CHANGES_REQUESTED** [review]

File: package.json [hunk 19]
```diff
@@ -61,6 +61,7 @@
     "react-router-dom": "^5.2.0",
     "react-scripts": "^5.0.0",
     "react-share": "^4.4.1",
+    "react-toastify": "^9.1.3",
```
Comment (Sutara-Skanda) [inline package.json:64]: use @capacitor/toast plugin instead react-toastify

File: src/components/DownloadChapterAndLesson.css [hunk 20]
```diff
@@ -0,0 +1,40 @@
+.downloadAndDeleteButton {
+  height: 35px;
+  width: 36px;
+  border-radius: 50%;
+  background-color: rgba(251, 250, 250);
+  cursor: "pointer";
+}
+
+.img {
```
Comment (Sutara-Skanda) [inline src/components/DownloadChapterAndLesson.css:9]: don't use generic names anywhere

File: src/components/DownloadChapterAndLesson.tsx [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+            <div className="spinner"></div>
+          </div>
+        </div>
+      ) : showIcon ? (
+        <div className="download">
+          <TfiDownload className="downloadButton" />
+        </div>
+      ) : (
+        <div
+          className="delete"
```
Comment (Sutara-Skanda) [inline src/components/DownloadChapterAndLesson.tsx:194]: same here

**Review by Sutara-Skanda — CHANGES_REQUESTED** [review]

**Review by Sutara-Skanda — APPROVED** [review]

**Review by bhanugummana — APPROVED** [review]

**Pull request merged**